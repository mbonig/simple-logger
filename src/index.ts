export * from './logger';
export { useLogger } from './useLogger';
