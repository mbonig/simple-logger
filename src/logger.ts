
export interface ILogger {
  log(message: string, ...args: any[]): void;
}

export class ConsoleLogger implements ILogger {
  log(message: string, ...args: any[]): void {
    console.log(message, args);
  }
}

export class LoggerFactory {
  public static getInstance(): ILogger {
    if (!this.logger) {
      this.logger = new ConsoleLogger();
    }
    return this.logger;
  }

  private static logger: ILogger;
}

