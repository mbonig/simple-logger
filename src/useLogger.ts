// @ts-ignore
import { around } from 'aspectjs';
import { LoggerFactory } from './logger';


function isPromise(p: any) {
  return p !== null &&
    typeof p === 'object' &&
    typeof p.then === 'function' &&
    typeof p.catch === 'function';
}

function getFunctionName(func: any) {
  return func.name ?? func.toString();
}

function serializeArgs(args: any) {
  if (typeof args === 'object') {
    try {
      return JSON.stringify(args, null, 2);
    } catch (err) {
      return args.toString();
    }
  }
  return args;

}


function logInputs(func: any, args: any) {
  const logger = LoggerFactory.getInstance();

  let serializedArgs: string;
  serializedArgs = serializeArgs(args);
  let functionName = getFunctionName(func);
  logger.log(`Executing ${functionName} with args: ${serializedArgs}`);
}

function logOutputs(func: any, returnValue: any) {
  const logger = LoggerFactory.getInstance();

  const functionName = getFunctionName(func);
  if (isPromise(returnValue)) {
    returnValue.then((val: any) => logger.log(`Function ${functionName} returned: ${serializeArgs(val)}`));
  } else {
    logger.log(`Function ${functionName} returned: ${serializeArgs(returnValue)}`);
  }

}


export function useLogger(func: any) {
  return around(func).add((invocation: any) => {
    const args = invocation.args;
    logInputs(func, args);
    const returnValue = invocation.proceed();
    logOutputs(func, returnValue);
    return returnValue;
  });
}
