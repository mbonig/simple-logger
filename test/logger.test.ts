import { useLogger } from '../src';
import { LoggerFactory } from '../src/logger';
import Mock = jest.Mock;


let mockLogger: Mock;

beforeEach(() => {
  jest.mock('../src/logger');
  mockLogger = jest.fn();
  LoggerFactory.getInstance = jest.fn().mockReturnValue({
    log: mockLogger,
  });
});
afterEach(() => {
  mockLogger.mockReset();
});

describe('Logs async things', () => {
  test('Logs async things', async () => {
    async function add(one: number, two: number) {
      return one + two;
    }

    await useLogger(add)(4, 5);

    expect(mockLogger.mock.calls[0]).toMatchSnapshot();
    expect(mockLogger.mock.calls[1]).toMatchSnapshot();
  });

  test('Logs async objects', async () => {
    async function add(one: number, two: number) {
      return { result: one + two };
    }

    await useLogger(add)(4, 5);

    expect(mockLogger.mock.calls[0]).toMatchSnapshot();
    expect(mockLogger.mock.calls[1]).toMatchSnapshot();
  });

});

describe('Logs sync things', () => {
  test('Logs things', () => {

    function add(one: number, two: number) {
      return one + two;
    }

    useLogger(add)(3, 4);

    expect(mockLogger.mock.calls[0]).toMatchSnapshot();
    expect(mockLogger.mock.calls[1]).toMatchSnapshot();
  });
  test('Logs things that are objects', () => {

    function add(one: number, two: number) {
      return { results: one + two };
    }

    useLogger(add)(3, 4);

    expect(mockLogger.mock.calls[0]).toMatchSnapshot();
    expect(mockLogger.mock.calls[1]).toMatchSnapshot();
  });

});

