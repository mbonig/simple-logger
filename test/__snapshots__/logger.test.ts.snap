// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs async things Logs async objects 1`] = `
Array [
  "Executing add with args: {
  \\"0\\": 4,
  \\"1\\": 5
}",
]
`;

exports[`Logs async things Logs async objects 2`] = `
Array [
  "Function add returned: {
  \\"result\\": 9
}",
]
`;

exports[`Logs async things Logs async things 1`] = `
Array [
  "Executing add with args: {
  \\"0\\": 4,
  \\"1\\": 5
}",
]
`;

exports[`Logs async things Logs async things 2`] = `
Array [
  "Function add returned: 9",
]
`;

exports[`Logs sync things Logs things 1`] = `
Array [
  "Executing add with args: {
  \\"0\\": 3,
  \\"1\\": 4
}",
]
`;

exports[`Logs sync things Logs things 2`] = `
Array [
  "Function add returned: 7",
]
`;

exports[`Logs sync things Logs things that are objects 1`] = `
Array [
  "Executing add with args: {
  \\"0\\": 3,
  \\"1\\": 4
}",
]
`;

exports[`Logs sync things Logs things that are objects 2`] = `
Array [
  "Function add returned: {
  \\"results\\": 7
}",
]
`;
